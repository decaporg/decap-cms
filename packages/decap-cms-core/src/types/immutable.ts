import { Map as ImmutableMap, List } from 'immutable';

export function isImmutableMap(value: unknown): value is ImmutableMap<string, unknown> {
  return ImmutableMap.isMap(value);
}

export function isImmutableList(value: unknown): value is List<unknown> {
  return List.isList(value);
}

export interface StaticallyTypedRecord<T> {
  get<K extends keyof T>(key: K, defaultValue?: T[K]): T[K];
  set<K extends keyof T, V extends T[K]>(key: K, value: V): StaticallyTypedRecord<T> & T;
  has<K extends keyof T>(key: K): boolean;
  delete<K extends keyof T>(key: K): StaticallyTypedRecord<T>;
  getIn<K1 extends keyof T, K2 extends keyof T[K1], V extends T[K1][K2]>(
    keys: [K1, K2],
    defaultValue?: V,
  ): T[K1][K2];
  getIn<
    K1 extends keyof T,
    K2 extends keyof T[K1],
    K3 extends keyof T[K1][K2],
    V extends T[K1][K2][K3],
  >(
    keys: [K1, K2, K3],
    defaultValue?: V,
  ): T[K1][K2][K3];
  getIn(keys: string[]): unknown;
  setIn<K1 extends keyof T, K2 extends keyof T[K1], V extends T[K1][K2]>(
    keys: [K1, K2],
    value: V,
  ): StaticallyTypedRecord<T>;
  setIn(keys: string[], value: unknown): StaticallyTypedRecord<T> & T;
  toJS(): T;
  isEmpty(): boolean;
  some<K extends keyof T>(predicate: (value: T[K], key: K, iter: this) => boolean): boolean;
  mapKeys<K extends keyof T, V>(mapFunc: (key: K, value: StaticallyTypedRecord<T>) => V): V[];
  find<K extends keyof T>(findFunc: (value: T[K]) => boolean): T[K];
  filter<K extends keyof T>(
    predicate: (value: T[K], key: K, iter: this) => boolean,
  ): StaticallyTypedRecord<T>;
  valueSeq<K extends keyof T>(): T[K][] & { toArray: () => T[K][] };
  map<K extends keyof T, V>(
    mapFunc: (value: T[K]) => V,
  ): StaticallyTypedRecord<{ [key: string]: V }>;
  keySeq<K extends keyof T>(): { toArray: () => K[] };
  withMutations(mutator: (mutable: StaticallyTypedRecord<T>) => unknown): StaticallyTypedRecord<T>;
}
