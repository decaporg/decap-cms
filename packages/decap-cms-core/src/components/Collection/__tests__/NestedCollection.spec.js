import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render, fireEvent } from '@testing-library/react';
import { fromJS } from 'immutable';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import ConnectedNestedCollection, {
  NestedCollection,
  getTreeData,
  walk,
  updateNode,
} from '../NestedCollection';

jest.mock('decap-cms-ui-default', () => {
  const actual = jest.requireActual('decap-cms-ui-default');
  return {
    ...actual,
    Icon: 'mocked-icon',
  };
});

const middlewares = [];
const mockStore = configureStore(middlewares);

function renderWithRedux(component, { store } = {}) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  return render(component, { wrapper: Wrapper });
}

describe('NestedCollection', () => {
  const collection = fromJS({
    name: 'pages',
    label: 'Pages',
    folder: 'src/pages',
    fields: [{ name: 'title', widget: 'string' }],
    nested: {
      subfolders: false,
    },
  });

  it('should render correctly with no entries', () => {
    const entries = fromJS([]);
    const { asFragment, getByTestId } = render(
      <MemoryRouter>
        <NestedCollection collection={collection} entries={entries} />
      </MemoryRouter>,
    );

    expect(getByTestId('/')).toHaveTextContent('Pages');
    expect(getByTestId('/')).toHaveAttribute('href', '/collections/pages');
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly with nested entries', () => {
    const entries = fromJS([
      { path: 'src/pages/index.md', data: { title: 'Root' } },
      { path: 'src/pages/a/index.md', data: { title: 'File 1' } },
      { path: 'src/pages/b/index.md', data: { title: 'File 2' } },
      { path: 'src/pages/a/a/index.md', data: { title: 'File 3' } },
      { path: 'src/pages/b/a/index.md', data: { title: 'File 4' } },
    ]);
    const { asFragment, getByTestId } = render(
      <MemoryRouter>
        <NestedCollection collection={collection} entries={entries} />
      </MemoryRouter>,
    );

    // expand the tree
    fireEvent.click(getByTestId('/'));

    expect(getByTestId('/a')).toHaveTextContent('File 1');
    expect(getByTestId('/a')).toHaveAttribute('href', '/collections/pages/filter/a');

    expect(getByTestId('/b')).toHaveTextContent('File 2');
    expect(getByTestId('/b')).toHaveAttribute('href', '/collections/pages/filter/b');

    expect(asFragment()).toMatchSnapshot();
  });

  it('should keep expanded nodes on re-render', () => {
    const entries = fromJS([
      { path: 'src/pages/index.md', data: { title: 'Root' } },
      { path: 'src/pages/a/index.md', data: { title: 'File 1' } },
      { path: 'src/pages/b/index.md', data: { title: 'File 2' } },
      { path: 'src/pages/a/a/index.md', data: { title: 'File 3' } },
      { path: 'src/pages/b/a/index.md', data: { title: 'File 4' } },
    ]);
    const { getByTestId, rerender } = render(
      <MemoryRouter>
        <NestedCollection collection={collection} entries={entries} />
      </MemoryRouter>,
    );

    fireEvent.click(getByTestId('/'));
    fireEvent.click(getByTestId('/a'));

    expect(getByTestId('/a')).toHaveTextContent('File 1');

    const newEntries = fromJS([
      { path: 'src/pages/index.md', data: { title: 'Root' } },
      { path: 'src/pages/a/index.md', data: { title: 'File 1' } },
      { path: 'src/pages/b/index.md', data: { title: 'File 2' } },
      { path: 'src/pages/a/a/index.md', data: { title: 'File 3' } },
      { path: 'src/pages/b/a/index.md', data: { title: 'File 4' } },
      { path: 'src/pages/c/index.md', data: { title: 'File 5' } },
      { path: 'src/pages/c/a/index.md', data: { title: 'File 6' } },
    ]);

    rerender(
      <MemoryRouter>
        <NestedCollection collection={collection} entries={newEntries} />
      </MemoryRouter>,
    );

    expect(getByTestId('/a')).toHaveTextContent('File 1');
  });

  it('should expand nodes based on filterTerm', () => {
    const entries = fromJS([
      { path: 'src/pages/index.md', data: { title: 'Root' } },
      { path: 'src/pages/a/index.md', data: { title: 'File 1' } },
      { path: 'src/pages/a/a/index.md', data: { title: 'File 2' } },
      { path: 'src/pages/a/a/a/index.md', data: { title: 'File 3' } },
    ]);

    const { getByTestId, queryByTestId, rerender } = render(
      <MemoryRouter>
        <NestedCollection collection={collection} entries={entries} />
      </MemoryRouter>,
    );

    expect(queryByTestId('/a/a')).toBeNull();

    rerender(
      <MemoryRouter>
        <NestedCollection collection={collection} entries={entries} filterTerm={'a/a'} />
      </MemoryRouter>,
    );

    expect(getByTestId('/a/a')).toHaveTextContent('File 2');
  });

  it('should ignore filterTerm once a user toggles an node', () => {
    const entries = fromJS([
      { path: 'src/pages/index.md', data: { title: 'Root' } },
      { path: 'src/pages/a/index.md', data: { title: 'File 1' } },
      { path: 'src/pages/a/a/index.md', data: { title: 'File 2' } },
      { path: 'src/pages/a/a/a/index.md', data: { title: 'File 3' } },
    ]);

    const { getByTestId, queryByTestId, rerender } = render(
      <MemoryRouter>
        <NestedCollection collection={collection} entries={entries} />
      </MemoryRouter>,
    );

    rerender(
      <MemoryRouter>
        <NestedCollection collection={collection} entries={entries} filterTerm={'a/a'} />
      </MemoryRouter>,
    );

    expect(getByTestId('/a/a')).toHaveTextContent('File 2');

    fireEvent.click(getByTestId('/a'));

    rerender(
      <MemoryRouter>
        <NestedCollection
          collection={collection}
          entries={fromJS(entries.toJS())}
          filterTerm={'a/a'}
        />
      </MemoryRouter>,
    );

    expect(queryByTestId('/a/a')).toBeNull();
  });

  it('should not collapse an unselected node when clicked', () => {
    const entries = fromJS([
      { path: 'src/pages/index.md', data: { title: 'Root' } },
      { path: 'src/pages/a/index.md', data: { title: 'File 1' } },
      { path: 'src/pages/a/a/index.md', data: { title: 'File 2' } },
      { path: 'src/pages/a/a/a/index.md', data: { title: 'File 3' } },
      { path: 'src/pages/a/a/a/a/index.md', data: { title: 'File 4' } },
    ]);

    const { getByTestId } = render(
      <MemoryRouter>
        <NestedCollection collection={collection} entries={entries} />
      </MemoryRouter>,
    );

    fireEvent.click(getByTestId('/'));
    fireEvent.click(getByTestId('/a'));
    fireEvent.click(getByTestId('/a/a'));

    expect(getByTestId('/a/a')).toHaveTextContent('File 2');
    fireEvent.click(getByTestId('/a'));
    expect(getByTestId('/a/a')).toHaveTextContent('File 2');
  });

  it('should collapse a selected node when clicked', () => {
    const entries = fromJS([
      { path: 'src/pages/index.md', data: { title: 'Root' } },
      { path: 'src/pages/a/index.md', data: { title: 'File 1' } },
      { path: 'src/pages/a/a/index.md', data: { title: 'File 2' } },
      { path: 'src/pages/a/a/a/index.md', data: { title: 'File 3' } },
      { path: 'src/pages/a/a/a/a/index.md', data: { title: 'File 4' } },
    ]);

    const { getByTestId, queryByTestId } = render(
      <MemoryRouter>
        <NestedCollection collection={collection} entries={entries} />
      </MemoryRouter>,
    );

    fireEvent.click(getByTestId('/'));
    fireEvent.click(getByTestId('/a'));
    fireEvent.click(getByTestId('/a/a'));

    expect(getByTestId('/a/a/a')).toHaveTextContent('File 3');
    fireEvent.click(getByTestId('/a/a'));
    expect(queryByTestId('/a/a/a')).toBeNull();
  });

  it('should render connected component', () => {
    const entriesArray = [
      { slug: 'index', path: 'src/pages/index.md', data: { title: 'Root' } },
      { slug: 'a/index', path: 'src/pages/a/index.md', data: { title: 'File 1' } },
      { slug: 'b/index', path: 'src/pages/b/index.md', data: { title: 'File 2' } },
      { slug: 'a/a/index', path: 'src/pages/a/a/index.md', data: { title: 'File 3' } },
      { slug: 'b/a/index', path: 'src/pages/b/a/index.md', data: { title: 'File 4' } },
    ];
    const entries = entriesArray.reduce(
      (acc, entry) => {
        acc.entities[`${collection.get('name')}.${entry.slug}`] = entry;
        acc.pages[collection.get('name')].ids.push(entry.slug);
        return acc;
      },
      { pages: { [collection.get('name')]: { ids: [] } }, entities: {} },
    );

    const store = mockStore({ entries: fromJS(entries) });

    const { asFragment, getByTestId } = renderWithRedux(
      <MemoryRouter>
        <ConnectedNestedCollection collection={collection} entries={entries} />
      </MemoryRouter>,
      { store },
    );

    // expand the root
    fireEvent.click(getByTestId('/'));

    expect(getByTestId('/a')).toHaveTextContent('File 1');
    expect(getByTestId('/a')).toHaveAttribute('href', '/collections/pages/filter/a');

    expect(getByTestId('/b')).toHaveTextContent('File 2');
    expect(getByTestId('/b')).toHaveAttribute('href', '/collections/pages/filter/b');

    expect(asFragment()).toMatchSnapshot();
  });

  describe('getTreeData', () => {
    it('should return nested tree data from entries', () => {
      const entries = fromJS([
        { path: 'src/pages/index.md', data: { title: 'Root' } },
        { path: 'src/pages/intro/index.md', data: { title: 'intro index' } },
        { path: 'src/pages/intro/category/index.md', data: { title: 'intro category index' } },
        { path: 'src/pages/compliance/index.md', data: { title: 'compliance index' } },
      ]);

      const treeData = getTreeData(collection, entries);

      expect(treeData).toEqual([
        {
          title: 'Pages',
          path: '/',
          isDir: true,
          isRoot: true,
          children: [
            {
              title: 'intro',
              path: '/intro',
              isDir: true,
              isRoot: false,
              children: [
                {
                  title: 'category',
                  path: '/intro/category',
                  isDir: true,
                  isRoot: false,
                  children: [
                    {
                      path: '/intro/category/index.md',
                      data: { title: 'intro category index' },
                      title: 'intro category index',
                      isDir: false,
                      isRoot: false,
                      children: [],
                    },
                  ],
                },
                {
                  path: '/intro/index.md',
                  data: { title: 'intro index' },
                  title: 'intro index',
                  isDir: false,
                  isRoot: false,
                  children: [],
                },
              ],
            },
            {
              title: 'compliance',
              path: '/compliance',
              isDir: true,
              isRoot: false,
              children: [
                {
                  path: '/compliance/index.md',
                  data: { title: 'compliance index' },
                  title: 'compliance index',
                  isDir: false,
                  isRoot: false,
                  children: [],
                },
              ],
            },
            {
              path: '/index.md',
              data: { title: 'Root' },
              title: 'Root',
              isDir: false,
              isRoot: false,
              children: [],
            },
          ],
        },
      ]);
    });

    it('should ignore collection summary', () => {
      const entries = fromJS([{ path: 'src/pages/index.md', data: { title: 'Root' } }]);

      const treeData = getTreeData(collection, entries);

      expect(treeData).toEqual([
        {
          title: 'Pages',
          path: '/',
          isDir: true,
          isRoot: true,
          children: [
            {
              path: '/index.md',
              data: { title: 'Root' },
              title: 'Root',
              isDir: false,
              isRoot: false,
              children: [],
            },
          ],
        },
      ]);
    });

    it('should use nested collection summary for title', () => {
      const entries = fromJS([{ path: 'src/pages/index.md', data: { title: 'Root' } }]);

      const treeData = getTreeData(
        collection.setIn(['nested', 'summary'], '{{filename}}'),
        entries,
      );

      expect(treeData).toEqual([
        {
          title: 'Pages',
          path: '/',
          isDir: true,
          isRoot: true,
          children: [
            {
              path: '/index.md',
              data: { title: 'Root' },
              title: 'index',
              isDir: false,
              isRoot: false,
              children: [],
            },
          ],
        },
      ]);
    });
  });

  describe('walk', () => {
    it('should visit every tree node', () => {
      const entries = fromJS([
        { path: 'src/pages/index.md', data: { title: 'Root' } },
        { path: 'src/pages/dir1/index.md', data: { title: 'Dir1 File' } },
        { path: 'src/pages/dir2/index.md', data: { title: 'Dir2 File' } },
      ]);

      const treeData = getTreeData(collection, entries);
      const callback = jest.fn();
      walk(treeData, callback);

      expect(callback).toHaveBeenCalledTimes(6);
      expect(callback).toHaveBeenCalledWith(expect.objectContaining({ path: '/' }));
      expect(callback).toHaveBeenCalledWith(expect.objectContaining({ path: '/index.md' }));
      expect(callback).toHaveBeenCalledWith(expect.objectContaining({ path: '/dir1' }));
      expect(callback).toHaveBeenCalledWith(expect.objectContaining({ path: '/dir2' }));
      expect(callback).toHaveBeenCalledWith(expect.objectContaining({ path: '/dir1/index.md' }));
      expect(callback).toHaveBeenCalledWith(expect.objectContaining({ path: '/dir2/index.md' }));
    });
  });

  describe('updateNode', () => {
    it('should update node', () => {
      const entries = fromJS([
        { path: 'src/pages/index.md', data: { title: 'Root' } },
        { path: 'src/pages/dir1/index.md', data: { title: 'Dir1 File' } },
        { path: 'src/pages/dir2/index.md', data: { title: 'Dir2 File' } },
      ]);

      const treeData = getTreeData(collection, entries);
      expect(treeData[0].children[0].children[0].expanded).toBeUndefined();

      const callback = jest.fn(node => ({ ...node, expanded: true }));
      const node = { path: '/dir1/index.md' };
      updateNode(treeData, node, callback);

      expect(callback).toHaveBeenCalledTimes(1);
      expect(callback).toHaveBeenCalledWith(node);
      expect(treeData[0].children[0].children[0].expanded).toEqual(true);
    });
  });
});
