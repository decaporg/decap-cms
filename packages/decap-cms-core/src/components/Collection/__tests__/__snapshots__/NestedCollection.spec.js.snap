// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NestedCollection should render connected component 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 14px;
  font-weight: 500;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px;
  padding-left: 18px;
  border-left: 2px solid #fff;
}

.emotion-0 mocked-icon {
  margin-right: 4px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-0:hover,
.emotion-0:active,
.emotion-0.sidebar-active {
  color: #3a69c7;
  background-color: #e8f5fe;
  border-left-color: #4863c6;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-4 {
  margin-right: 4px;
}

.emotion-6 {
  position: relative;
  top: 2px;
  color: #fff;
  width: 0;
  height: 0;
  border: 5px solid transparent;
  border-radius: 2px;
  border-top: 6px solid currentColor;
  border-bottom: 0;
  color: currentColor;
}

<a
    aria-current="page"
    class="emotion-0 emotion-1 sidebar-active"
    data-testid="/"
    depth="0"
    href="/collections/pages"
  >
    <mocked-icon
      type="write"
    />
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="emotion-4 emotion-5"
      >
        Pages
      </div>
      <div
        class="emotion-6 emotion-7"
      />
    </div>
  </a>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 14px;
  font-weight: 500;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px;
  padding-left: 34px;
  border-left: 2px solid #fff;
}

.emotion-0 mocked-icon {
  margin-right: 4px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-0:hover,
.emotion-0:active,
.emotion-0.sidebar-active {
  color: #3a69c7;
  background-color: #e8f5fe;
  border-left-color: #4863c6;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-4 {
  margin-right: 4px;
}

.emotion-6 {
  position: relative;
  top: 2px;
  color: #fff;
  width: 0;
  height: 0;
  border: 5px solid transparent;
  border-radius: 2px;
  border-left: 6px solid currentColor;
  border-right: 0;
  color: currentColor;
  left: 2px;
}

<a
    class="emotion-0 emotion-1"
    data-testid="/a"
    depth="1"
    href="/collections/pages/filter/a"
  >
    <mocked-icon
      type="write"
    />
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="emotion-4 emotion-5"
      >
        File 1
      </div>
      <div
        class="emotion-6 emotion-7"
      />
    </div>
  </a>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 14px;
  font-weight: 500;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px;
  padding-left: 34px;
  border-left: 2px solid #fff;
}

.emotion-0 mocked-icon {
  margin-right: 4px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-0:hover,
.emotion-0:active,
.emotion-0.sidebar-active {
  color: #3a69c7;
  background-color: #e8f5fe;
  border-left-color: #4863c6;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-4 {
  margin-right: 4px;
}

.emotion-6 {
  position: relative;
  top: 2px;
  color: #fff;
  width: 0;
  height: 0;
  border: 5px solid transparent;
  border-radius: 2px;
  border-left: 6px solid currentColor;
  border-right: 0;
  color: currentColor;
  left: 2px;
}

<a
    class="emotion-0 emotion-1"
    data-testid="/b"
    depth="1"
    href="/collections/pages/filter/b"
  >
    <mocked-icon
      type="write"
    />
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="emotion-4 emotion-5"
      >
        File 2
      </div>
      <div
        class="emotion-6 emotion-7"
      />
    </div>
  </a>
</DocumentFragment>
`;

exports[`NestedCollection should render correctly with nested entries 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 14px;
  font-weight: 500;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px;
  padding-left: 18px;
  border-left: 2px solid #fff;
}

.emotion-0 mocked-icon {
  margin-right: 4px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-0:hover,
.emotion-0:active,
.emotion-0.sidebar-active {
  color: #3a69c7;
  background-color: #e8f5fe;
  border-left-color: #4863c6;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-4 {
  margin-right: 4px;
}

.emotion-6 {
  position: relative;
  top: 2px;
  color: #fff;
  width: 0;
  height: 0;
  border: 5px solid transparent;
  border-radius: 2px;
  border-top: 6px solid currentColor;
  border-bottom: 0;
  color: currentColor;
}

<a
    aria-current="page"
    class="emotion-0 emotion-1 sidebar-active"
    data-testid="/"
    depth="0"
    href="/collections/pages"
  >
    <mocked-icon
      type="write"
    />
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="emotion-4 emotion-5"
      >
        Pages
      </div>
      <div
        class="emotion-6 emotion-7"
      />
    </div>
  </a>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 14px;
  font-weight: 500;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px;
  padding-left: 34px;
  border-left: 2px solid #fff;
}

.emotion-0 mocked-icon {
  margin-right: 4px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-0:hover,
.emotion-0:active,
.emotion-0.sidebar-active {
  color: #3a69c7;
  background-color: #e8f5fe;
  border-left-color: #4863c6;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-4 {
  margin-right: 4px;
}

.emotion-6 {
  position: relative;
  top: 2px;
  color: #fff;
  width: 0;
  height: 0;
  border: 5px solid transparent;
  border-radius: 2px;
  border-left: 6px solid currentColor;
  border-right: 0;
  color: currentColor;
  left: 2px;
}

<a
    class="emotion-0 emotion-1"
    data-testid="/a"
    depth="1"
    href="/collections/pages/filter/a"
  >
    <mocked-icon
      type="write"
    />
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="emotion-4 emotion-5"
      >
        File 1
      </div>
      <div
        class="emotion-6 emotion-7"
      />
    </div>
  </a>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 14px;
  font-weight: 500;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px;
  padding-left: 34px;
  border-left: 2px solid #fff;
}

.emotion-0 mocked-icon {
  margin-right: 4px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-0:hover,
.emotion-0:active,
.emotion-0.sidebar-active {
  color: #3a69c7;
  background-color: #e8f5fe;
  border-left-color: #4863c6;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-4 {
  margin-right: 4px;
}

.emotion-6 {
  position: relative;
  top: 2px;
  color: #fff;
  width: 0;
  height: 0;
  border: 5px solid transparent;
  border-radius: 2px;
  border-left: 6px solid currentColor;
  border-right: 0;
  color: currentColor;
  left: 2px;
}

<a
    class="emotion-0 emotion-1"
    data-testid="/b"
    depth="1"
    href="/collections/pages/filter/b"
  >
    <mocked-icon
      type="write"
    />
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="emotion-4 emotion-5"
      >
        File 2
      </div>
      <div
        class="emotion-6 emotion-7"
      />
    </div>
  </a>
</DocumentFragment>
`;

exports[`NestedCollection should render correctly with no entries 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 14px;
  font-weight: 500;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px;
  padding-left: 18px;
  border-left: 2px solid #fff;
}

.emotion-0 mocked-icon {
  margin-right: 4px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-0:hover,
.emotion-0:active,
.emotion-0.sidebar-active {
  color: #3a69c7;
  background-color: #e8f5fe;
  border-left-color: #4863c6;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-4 {
  margin-right: 4px;
}

.emotion-6 {
  position: relative;
  top: 2px;
  color: #fff;
  width: 0;
  height: 0;
  border: 5px solid transparent;
  border-radius: 2px;
  border-left: 6px solid currentColor;
  border-right: 0;
  color: currentColor;
  left: 2px;
}

<a
    class="emotion-0 emotion-1"
    data-testid="/"
    depth="0"
    href="/collections/pages"
  >
    <mocked-icon
      type="write"
    />
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="emotion-4 emotion-5"
      >
        Pages
      </div>
      <div
        class="emotion-6 emotion-7"
      />
    </div>
  </a>
</DocumentFragment>
`;
